##load project related functions
source('function.r')

##setup function
parseCommandArgs <- function (evaluate = TRUE) 
{
  arglist <- list()
  args <- commandArgs()
  i <- which(args == "--args")
  if (length(i) == 0 || length(args) < 1) 
    return(invisible())
  args <- args[(i + 1):length(args)]
  for (i in seq(1, length(args), by = 2)) {
    value <- NA
    tryCatch(value <- as.double(args[i + 1]), warning = function(e) {
    })
    if (is.na(value)) {
      value <- args[i + 1]
      if (substr(value, 1, 2) == "c(") 
        value <- eval(parse(text = args[i + 1]))
    }
    if (evaluate) 
      assign(args[i], value, inherits = TRUE)
    arglist[[length(arglist) + 1]] <- value
    names(arglist)[length(arglist)] <- args[i]
  }
  return(arglist)
}

#need to take in parameter
## passed in from the command prompt.
parseCommandArgs()
## Will disply the default values on the first run,
print(seed)
print(n_rep)
print(f)
print(n_family) #number of family
print(p_dis) #prevelance

## ... your simualtion code
##the founder's haplotypes
#import the haplotypes generated by cosi
haplotype <- read.table("out_100k_10k_1kb.hap-1", header=F)
colnames(haplotype) <- c("HAP", "CHROM", paste("SNP", 1:(ncol(haplotype)-2), sep=""))
snp <-read.table("out_100k_10k_1kb.pos-1", header=T)
#make allele 1 is the minor allele, 2 is the common allele
temp.idx <- snp$FREQ1 > snp$FREQ2
temp.freq <- snp$FREQ2
snp$FREQ2[temp.idx] <- snp$FREQ1[temp.idx]
snp$FREQ1[temp.idx] <- temp.freq[temp.idx]
#also change the genotype file
haplotype[,which(temp.idx==T)+2] <- 3 - haplotype[,which(temp.idx==T)+2]

#allele frequency
nrow(snp) #total number of snp
sum(snp$FREQ1 < 0.05) # number of snp with f < 0.05
sum(snp$FREQ1 < 0.01) # number of snp with f < 0.05
sum(snp$FREQ1 == 0.0001) # number of snp with f < 0.05

##assign risk variants and the corresponding effect size (proportional to allele frequency)
n_haplo <- 10000
n_snp <- ncol(haplotype)-2
prevalence <- 0.05
b0_sqrt <- sqrt(prevalence)   #baseline
temp.idx <- sort(c(sample(which((snp$FREQ1 < 0.05)==T & snp$FREQ1 != 1/n_haplo ), 5), sample(which(snp$FREQ1 == 1/n_haplo), 5))) #randomly pick 5 variants with f < 0.05 and 5 singletons as casual 
b <- rep(1, length=(ncol(haplotype)-2))
b[temp.idx] <- abs(log10(snp$FREQ1[temp.idx])) #effect size is log10 of its allele frequency
#calculate the haplotype variants p(A|h)
haplotype.risk <- apply(2-haplotype[, -c(1:2)], 1, function(x) prod(b^x)*b0_sqrt)
#calculate the probability of drawing given the affected status of founder p(h|A)
# haplotype.draw_prob <- (haplotype.risk %o% haplotype.risk)/n_haplo^2/prevalence
mean(b) #mean relative risk
mean(apply(2-haplotype[, which((snp$FREQ1 < 0.01)==T)+2], 1, sum)) #carrier haplotype frequency


rep.idx <<- 1
sim_result <- replicate(n_rep, {
  print(rep.idx)
  rep.idx <<- rep.idx + 1
  family_generated <- gene_family(n=n_family) 
  family.test(data=family_generated, f=f)
})
## Write out your results to a csv file
write.csv(data.frame(seed=seed, f=f, finail.test.stat=sim_result[1,], sum_e=sim_result[2,], se=sim_result[3,], 
                     p.value=sim_result[4,], n_info_family=sim_result[5,]),
          paste("res_",f,"_",n_family,"_",seed,".csv",sep=""), row.names=FALSE)
## R.miSniam
## End(Not run)
## Not run: