## ... your simualtion codek
source("/net/frodo/home/khlin/p3/function.r")

#need to take in parameter
## passed in from the command prompt.
parseCommandArgs()
## Will disply the default values on the first run,
print(seed)
print(n_rep)
print(r)
print(n_family) #number of family
print(p_dis) #prevelance
# print(risk.variant)
print(family_strct)
print(f) #snps with f to consider

##1. read in cosi haplotype
#import the haplotypes generated by cosi
haplotype <- read.table("out_100k_10k_1kb.hap-1", header=F)
colnames(haplotype) <- c("HAP", "CHROM", paste("SNP", 1:(ncol(haplotype)-2), sep=""))
snp <-read.table("out_100k_10k_1kb.pos-1", header=T)
#make allele 1 is the minor allele, 2 is the common allele
temp.idx <- snp$FREQ1 > snp$FREQ2
temp.freq <- snp$FREQ2
snp$FREQ2[temp.idx] <- snp$FREQ1[temp.idx]
snp$FREQ1[temp.idx] <- temp.freq[temp.idx]
#also change the genotype file
haplotype[,which(temp.idx==T)+2] <- 3 - haplotype[,which(temp.idx==T)+2]
n_snp <- nrow(snp)

##2. gives structure and 
family_strct_ped <- select_family_strct(family_strct)

##3.load library for pedgene and fbat_ext
.libPaths(c("/net/frodo/home/khlin/R/x86_64-pc-linux-gnu-library/3.2/", .libPaths()))
library(pedgene)
library(methods)
library(SKAT)
n_cc <- length(family_strct_ped$person)/2*n_family

##4. effect size
# risk.variant.id <- c(5,15,16,19,20,27,31,35,43,49)
(n_snp_f <- sum(snp$FREQ1<f))
(n_snp_f_causal <- round(n_snp_f*0.1)) #10 percent of snp with maf < f are causal
# risk.variant.id <<- sort(sample(which(0.0001<snp$FREQ1 & snp$FREQ1<f), round(n_snp_f_causal)))
# risk.haplo.id <- which(apply(2-as.matrix(haplotype[, risk.variant.id+2]), 1, sum)>0)
risk.variant.id <- switch(as.character(f),
                          "0.01"=c(4,37,38),
                          "0.05"=c(12,20,37,44),
                          "0.2"=c(1,4,22,26)
)
risk.haplo.f <<- mean(apply(2-as.matrix(haplotype[, risk.variant.id+2]), 1, sum)>0) #carrier haplotype frequency
cat("risk.haplo.f = ", risk.haplo.f, "\n")
Beta <- rep(0, length.out=50)
Beta[risk.variant.id] <- log(r) #effect size of OR=2

##5 .create genes.txt weight.txt variant_pass.txt for fb-skat
write(paste("geneA 1", n_snp_f_causal), paste0("genes_",seed,".txt"))
write(rep("1", n_snp_f_causal), paste0("weight_",seed,".txt"), sep="\n")
write(rep("1", n_snp_f_causal), paste0("variant_pass_",seed,".txt"), sep="\n")

sim_result <- list()
for(i in 1:n_rep) {
  print(i)
  sim.fail <- F
  
  family_generated <<- gene_family_pe(family_strct=family_strct_ped, n_family=n_family, p_dis=p_dis, Beta=Beta)
  
  #remove the founder from the data
  family_generated_c <- family_generated
  temp.idx <- which(family_generated$data_family$person %in% 1:2) #take out founders
  family_generated_c$data_family <- family_generated$data_family[-temp.idx, ]
  family_generated_c$tran_vec <- family_generated$tran_vec[-temp.idx, ]
  
  ##test based on new imputation framework
  result.trap.3c.noimpute <- family.test(data=family_generated, f=risk.variant.id, nofounderphenotype=T)$p.value
  result.trap.3c.impute.founder.pop.f <- family.test.nofounder.impute(data=family_generated_c, f=risk.variant.id, sample.f=F, pop.f.off=0)
  result.trap.3c.impute.founder.pop.f.off.10 <- family.test.nofounder.impute(data=family_generated_c, f=risk.variant.id, sample.f=F, pop.f.off=10)
  result.trap.3c.impute.founder.pop.f.off.20 <- family.test.nofounder.impute(data=family_generated_c, f=risk.variant.id, sample.f=F, pop.f.off=20)
  result.trap.3c.impute.founder.pop.f.off.50 <- family.test.nofounder.impute(data=family_generated_c, f=risk.variant.id, sample.f=F, pop.f.off=50)
  result.trap.3c.impute.founder.pop.f.off.n10 <- family.test.nofounder.impute(data=family_generated_c, f=risk.variant.id, sample.f=F, pop.f.off=-10)
  result.trap.3c.impute.founder.pop.f.off.n20 <- family.test.nofounder.impute(data=family_generated_c, f=risk.variant.id, sample.f=F, pop.f.off=-20)
  result.trap.3c.impute.founder.pop.f.off.n50 <- family.test.nofounder.impute(data=family_generated_c, f=risk.variant.id, sample.f=F, pop.f.off=-50)
  result.trap.3c.impute.founder.sample.f <- family.test.nofounder.impute(data=family_generated_c, f=risk.variant.id, sample.f=T)
  
  #skip return when errors occured
  if(sim.fail==F) {
    #only report p.value
    sim_result[[i]] <- data.frame(result.trap.3c.noimpute, result.trap.3c.impute.founder.pop.f,
                                  result.trap.3c.impute.founder.pop.f.off.10,
                                  result.trap.3c.impute.founder.pop.f.off.20, result.trap.3c.impute.founder.pop.f.off.50,
                                  result.trap.3c.impute.founder.pop.f.off.n10,
                                  result.trap.3c.impute.founder.pop.f.off.n20, result.trap.3c.impute.founder.pop.f.off.n50,
                                  result.trap.3c.impute.founder.sample.f) 
  } else {
    warning("error occured!!")
    next
  }
}
##remove junk files produced by fbskat
system(paste("rm results_",seed,"*.txt", sep=""))
system(paste("rm mendelian_errors_",seed,"*.txt", sep=""))
system(paste("rm data_",seed,"*.ped", sep=""))
system(paste("rm genes_",seed,"*.txt", sep=""))
system(paste("rm weight_",seed,"*.txt", sep=""))
system(paste("rm variant_pass_",seed,"*.txt", sep=""))
## Write out your results to a csv file
result.df <- do.call(rbind, sim_result)
result.df <- cbind(seed,f,r,p_dis,risk.variant.id=paste(c(risk.variant.id), collapse = "_"),risk.haplo.f,n_family,family_strct,result.df)
write.csv(result.df, paste("res_",seed,"_",r,"_",f,"_",n_family,".csv",sep=""), row.names=FALSE)
## R.miSniam
## End(Not run)
## Not run:
