## ... your simualtion codek
source("/net/frodo/home/khlin/p4/function.r")

#need to take in parameter
## passed in from the command prompt.
parseCommandArgs()
## Will disply the default values on the first run,
print(seed)
print(n_rep)
print(r)
print(n_family) #number of family
print(trait_mean) #trait mean
print(dis_cutoff)
# print(risk.variant)
print(family_strct)
print(f) #snps with f to consider
(exact_affected <- eval(parse(text=exact_affected)))

##1. read in cosi haplotype
#import the haplotypes generated by cosi
haplotype <- read.table("out_100k_10k_1kb.hap-1", header=F)
colnames(haplotype) <- c("HAP", "CHROM", paste("SNP", 1:(ncol(haplotype)-2), sep=""))
snp <-read.table("out_100k_10k_1kb.pos-1", header=T)
#make allele 1 is the minor allele, 2 is the common allele
temp.idx <- snp$FREQ1 > snp$FREQ2
temp.freq <- snp$FREQ2
snp$FREQ2[temp.idx] <- snp$FREQ1[temp.idx]
snp$FREQ1[temp.idx] <- temp.freq[temp.idx]
#also change the genotype file
haplotype[,which(temp.idx==T)+2] <- 3 - haplotype[,which(temp.idx==T)+2]
n_snp <- nrow(snp)

##2. gives structure and 
family_strct_ped <- select_family_strct(family_strct)

##3.load library for pedgene and fbat_ext
.libPaths(c("/net/frodo/home/khlin/R/x86_64-pc-linux-gnu-library/3.2/", .libPaths()))
library(pedgene)
library(methods)
library(SKAT)
n_cc <- length(family_strct_ped$person)/2*n_family

##4. effect size
# risk.variant.id <- c(5,15,16,19,20,27,31,35,43,49)
(n_snp_f <- sum(snp$FREQ1<f))
(n_snp_f_causal <- round(n_snp_f*0.1)) #10 percent of snp with maf < f are causal
# risk.variant.id <<- sort(sample(which(0.0001<snp$FREQ1 & snp$FREQ1<f), round(n_snp_f_causal)))
# risk.haplo.id <- which(apply(2-as.matrix(haplotype[, risk.variant.id+2]), 1, sum)>0)
risk.variant.id <- switch(as.character(f),
                          "0.01"=c(4,37,38),
                          "0.05"=c(12,20,37,44),
                          "0.2"=c(1,4,22,26)
)
risk.haplo.f <<- mean(apply(2-as.matrix(haplotype[, risk.variant.id+2]), 1, sum)>0) #carrier haplotype frequency
cat("risk.haplo.f = ", risk.haplo.f, "\n")
Beta <- rep(0, length.out=50)
Beta[risk.variant.id] <- r #effect size in linear model

##5 .create genes.txt weight.txt variant_pass.txt for fb-skat
write(paste("geneA 1", n_snp_f_causal), paste0("genes_",seed,".txt"))
write(rep("1", n_snp_f_causal), paste0("weight_",seed,".txt"), sep="\n")
write(rep("1", n_snp_f_causal), paste0("variant_pass_",seed,".txt"), sep="\n")

sim_result <- list()
for(i in 1:n_rep) {
  print(i)
  sim.fail <- F
  
  #simulation for two-generation families
  family_generated <<- gene_family_pe(family_strct=family_strct_ped, n_family=n_family, trait_mean=trait_mean, Beta=Beta, dis_cutoff = dis_cutoff, exact_affected = exact_affected) 
  #add pseudo null variant since FB-SKAT can only work with 2+ in a gene
  family_generated_diploid <- hap2dip(data=family_generated, risk.variant.id=risk.variant.id, save.file=T)
  
  #run tests 
  result.trap.ind <- family.test(data=family_generated, f=risk.variant.id, summary.stat="3")$p.value
  result.trap.rank.ind <- family.rank.test(data=family_generated, f=risk.variant.id, summary.stat="3")$p.value
  sim.fail <- tryCatch({  
    result.pedgene <- pedgene(ped=family_generated_diploid$ped, geno=family_generated_diploid$geno)
    result.pedgene.vc <- result.pedgene$pgdf$pval.kernel
    result.pedgene.burden <- result.pedgene$pgdf$pval.burden
    F
  }, 
  error = function(e) return(T) 
  )
  
#   system(paste("/net/frodo/home/khlin/p3/FB-SKAT/FB-SKAT_v2.1 data_",seed,".ped variant_pass_",seed,".txt genes_",seed,".txt weight_",seed,".txt results_", seed, "_ mendelian_errors_",seed,"_ 10000 1 0.01 0 0", sep=""))
#   system(paste("/net/frodo/home/khlin/p3/FB-SKAT/FB-SKAT_v2.1 data_",seed,".ped variant_pass_",seed,".txt genes_",seed,".txt weight_",seed,".txt results_", seed, "_ mendelian_errors_",seed,"_ 10000 1 0.01 1 0", sep=""), ignore.stdout = TRUE)
#   x <- read.table(paste("results_", seed,"_1.000000_0.010000_0.txt", sep="")) #variance component
#   result.fbskat.vc <- 1-pchisq(x[,4],x[,5])
#   x <- read.table(paste("results_", seed,"_1.000000_0.010000_1.txt", sep="")) #burden
#   result.fbskat.burden=1-pchisq(x[,4],x[,5])
  
  #case-control use burden test with equal weight
  data.cc <- gene_case_control_pe(n_case=n_cc, n_control=n_cc, trait_mean=trait_mean, Beta=Beta, dis_cutoff = dis_cutoff)
  data.cc.diploid <- hap2dip(data=list(data_family=data.cc), risk.variant.id=risk.variant.id, save.file=F)
  obj <- SKAT_Null_Model(data.cc.diploid$ped$trait ~ 1, out_type="C")
  result.pop <- SKAT(as.matrix(data.cc.diploid$geno[, -c(1:2)]), obj, weights.beta = c(1,1), r.corr = 1)$p.value
  

  #centered_trait
  family_generated_centered <- family_generated
  family_generated_centered$data_family[,"affect"] <- family_generated_centered$data_family[,"affect"] - trait_mean
  #add pseudo null variant since FB-SKAT can only work with 2+ in a gene
  family_generated_centered_diploid <- hap2dip(data=family_generated_centered, risk.variant.id=risk.variant.id, save.file=T)
  
  #run tests 
  result.trap.ind.centered <- family.test(data=family_generated_centered, f=risk.variant.id, summary.stat="3")$p.value
  result.trap.rank.ind.centered <- family.rank.test(data=family_generated_centered, f=risk.variant.id, summary.stat="3")$p.value

  sim.fail <- tryCatch({  
    result.pedgene <- pedgene(ped=family_generated_centered_diploid$ped, geno=family_generated_centered_diploid$geno)
    result.pedgene.vc.centered <- result.pedgene$pgdf$pval.kernel
    result.pedgene.burden.centered <- result.pedgene$pgdf$pval.burden
    F
  }, 
  error = function(e) return(T) 
  )
  
#   system(paste("/net/frodo/home/khlin/p3/FB-SKAT/FB-SKAT_v2.1 data_",seed,".ped variant_pass_",seed,".txt genes_",seed,".txt weight_",seed,".txt results_", seed, "_ mendelian_errors_",seed,"_ 10000 1 0.01 0 0", sep=""))
#   system(paste("/net/frodo/home/khlin/p3/FB-SKAT/FB-SKAT_v2.1 data_",seed,".ped variant_pass_",seed,".txt genes_",seed,".txt weight_",seed,".txt results_", seed, "_ mendelian_errors_",seed,"_ 10000 1 0.01 1 0", sep=""), ignore.stdout = TRUE)
#   x <- read.table(paste("results_", seed,"_1.000000_0.010000_0.txt", sep="")) #variance component
#   result.fbskat.vc.centered <- 1-pchisq(x[,4],x[,5])
#   x <- read.table(paste("results_", seed,"_1.000000_0.010000_1.txt", sep="")) #burden
#   result.fbskat.burden.centered <- 1-pchisq(x[,4],x[,5])
  
  #case-control use burden test with equal weight
  data.cc_centered <- data.cc
  data.cc_centered$affect <- data.cc_centered$affect - trait_mean 
  
  data.cc_centered.diploid <- hap2dip(data=list(data_family=data.cc_centered), risk.variant.id=risk.variant.id, save.file=F)
  obj <- SKAT_Null_Model(data.cc_centered.diploid$ped$trait ~ 1, out_type="C")
  result.pop.centered <- SKAT(as.matrix(data.cc_centered.diploid$geno[, -c(1:2)]), obj, weights.beta = c(1,1), r.corr = 1)$p.value
  
  #skip return when errors occured
  if(sim.fail==F) {
    #only report p.value
    sim_result[[i]] <- data.frame(result.trap.ind, result.trap.rank.ind,
                                  result.pedgene.vc, result.pedgene.burden,
                                  # result.fbskat.vc, result.fbskat.burden,
                                  result.pop,
                                  result.trap.ind.centered, result.trap.rank.ind.centered,
                                  result.pedgene.vc.centered, result.pedgene.burden.centered,
                                  # result.fbskat.vc.centered, result.fbskat.burden.centered,
                                  result.pop.centered) 
  } else {
    warning("error occured!!")
    next
  }
}
##remove junk files produced by fbskat
system(paste("rm results_",seed,"*.txt", sep=""))
system(paste("rm mendelian_errors_",seed,"*.txt", sep=""))
system(paste("rm data_",seed,"*.ped", sep=""))
system(paste("rm genes_",seed,"*.txt", sep=""))
system(paste("rm weight_",seed,"*.txt", sep=""))
system(paste("rm variant_pass_",seed,"*.txt", sep=""))
## Write out your results to a csv file
result.df <- do.call(rbind, sim_result)
result.df <- cbind(seed,f,r,trait_mean,dis_cutoff,risk.variant.id=paste(c(risk.variant.id), collapse = "_"),risk.haplo.f,n_family,family_strct,result.df,exact_affected)
write.csv(result.df, paste("res_",seed,"_",r,"_",f,"_",n_family,".csv",sep=""), row.names=FALSE)
## R.miSniam
## End(Not run)
## Not run:
